ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/ExplicitDisclosureIT.scala

```diff
@@ -0,0 +1,421 @@
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    val contractKey = ledger.nextKeyId()
+    for {
+      _ <- ledger.create(owner, Delegated(owner, contractKey))
+      _ <- ledger.create(owner, Delegation(owner, delegate))
+      flats <- ledger.flatTransactions(owner)
+      trees <- ledger.transactionTrees(owner)
+      someTransactionId = flats.head.transactionId
+      flatById <- ledger.flatTransactionById(someTransactionId, owner)
+      treeById <- ledger.transactionTreeById(someTransactionId, owner)
+    } yield {
+      assertLength("flatTransactions", 2, flats)
+      assertLength("transactionTrees", 2, trees)
+      assert(
+        flats.map(createdEvents).forall(_.forall(_.metadata.isDefined)),
+        "Metadata is empty for flatTransactions",
+      )
+      assert(
+        trees.map(createdEvents).forall(_.forall(_.metadata.isDefined)),
+        "Metadata is empty for transactionTrees",
+      )
+      assert(
+        createdEvents(flatById).forall(_.metadata.isDefined),
+        "Metadata is empty for flatTransactionById",
+      )
+      assert(
+        createdEvents(treeById).forall(_.metadata.isDefined),
+        "Metadata is empty for transactionTreeById",
+      )
+    }
+  })
+
+  test(
+    "EDArchivedDisclosedContracts",
+    "The ledger rejects archived disclosed contracts",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    for {
+      testContext <- initializeTest(ledger, owner, delegate)
+
+      // Archive the disclosed contract
+      _ <- ledger.exercise(owner, testContext.delegatedCid.exerciseArchive())
+
+      // Exercise the choice using the now inactive disclosed contract
+      exerciseError <- testContext
+        .exerciseFetchDelegated(testContext.disclosedContract)
+        .mustFail("the contract is already archived")
+    } yield {
+      assertGrpcError(
+        exerciseError,
+        LedgerApiErrors.ConsistencyErrors.ContractNotFound,
+        None,
+        checkDefiniteAnswerMetadata = true,
+      )
+    }
+  })
+
+  test(
+    "EDInconsistentDisclosedContracts",
+    "The ledger rejects disclosed contracts with inconsistent metadata",
+    allocate(Parties(2)),
+    enabled = _.explicitDisclosure,
+  )(implicit ec => { case Participants(Participant(ledger, owner, delegate)) =>
+    for {
+      testContext <- initializeTest(ledger, owner, delegate)
+
+      // Exercise a choice using invalid explicit disclosure (mismatching contract id)
+      errorMismatchingContractId <- testContext
+        .exerciseFetchDelegated(
+          testContext.disclosedContract
+            .update(_.contractId := ContractId.V1.assertFromString("00" + "00" * 32).coid)
+        )
+        .mustFail("using a mismatching contract id")
+
+      // Exercise a choice using invalid explicit disclosure (bad contract key)
```

Comment:
wouldn’t this blow up with bad payload rather than bad key? I expect bad key hash if the hash does not match the key but the payload does match the actual contract.

Reply:
You're right. Adapting it to actually test the key hash in the metadata.